keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,816 @@
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+When a cluster has multiple apiservers at mixed versions (such as during an
+upgrade or downgrate), not every apiserver can serve every resource at every
+version.
+
+To fix this, we will add a filter to the handler chain in the aggregator which
+proxies clients to an apiserver that is capable of handling their request.
+
+## Motivation
+
+When an upgrade or downgrade is performed on a cluster, for some period of time
+the apiservers are at differing versions and are able to serve different sets of
+built-in resources (different groups, versions, and resources are all possible).
+
+In an ideal world, clients would be able to know about the entire set of
+available resources and perform operations on those resources without regard to
+which apiserver they happened to connect to. Currently this is not the case.
+
+Today, these things potentially differ:
+* Resources available somewhere in the cluster
+* Resources known by a client (i.e. read from discovery from some apiserver)
+* Resources that can be actuated by a client
+
+This can have serious consequences, such as namespace deletion being blocked
+incorrectly or objects being garbage collected mistakenly.
+
+### Goals
+
+* Ensure discovery reports the same set of resources everywhere (not just group
+  versions, as it does today)
+* Proxy client traffic to an apiserver that can service the request
```

Comment:
Have you considered instead disabling resources which are not served by all API servers?

Reply:
Then you end up with an intersection of the resources of both sets of apiservers, which is still going to lead to wonky behavior from the GC and NLC controllers.